File: Practica_Iterator1/book_collection.py
# book_collection.py

# Importamos los iteradores concretos que esta colección puede proporcionar.
# Esto demuestra que el Aggregate es responsable de crear (instanciar)
# los iteradores adecuados para sí mismo.
from book_iterator import BookIterator
from reverse_book_iterator import ReverseBookIterator
from author_alphabetical_iterator import AuthorAlphabeticalIterator

# La clase BookCollection es el 'Concrete Aggregate' (Agregado Concreto).
# Es la colección de objetos que queremos recorrer.
# Su principal responsabilidad es proporcionar una interfaz para crear objetos Iterator.
class BookCollection:
    def __init__(self):
        # La representación interna de la colección (una lista en este caso).
        # El patrón Iterator nos permite recorrerla sin exponer esta lista directamente.
        self._books = []

    def add_book(self, book):
        # Método para añadir elementos a la colección.
        self._books.append(book)

    # Este método mágico __iter__ es crucial para el patrón Iterator en Python.
    # Es la implementación del método 'createIterator()' del patrón.
    # Permite que las instancias de BookCollection sean iterables directamente
    # en bucles 'for', proporcionando un iterador por defecto (el normal).
    def __iter__(self):
        """
        Retorna el iterador por defecto (normal) para la colección.
        Esta es la implementación del método 'createIterator()' del patrón Iterator.
        """
        # Devolvemos una nueva instancia de un Concrete Iterator.
        return BookIterator(self._books)

    def get_reverse_iterator(self):
        """
        Retorna un iterador inverso para la colección.
        Esto demuestra que un Aggregate puede ofrecer múltiples tipos de iteradores,
        cada uno para una forma diferente de recorrido.
        """
        return ReverseBookIterator(self._books)

    def get_author_alphabetical_iterator(self):
        """
        Retorna un iterador que ordena los libros por autor alfabéticamente.
        Otro ejemplo de cómo el Aggregate puede proporcionar iteradores especializados.
        """
        return AuthorAlphabeticalIterator(self._books)

    def __len__(self):
        # Método para obtener la longitud de la colección, útil para el iterador inverso.
        return len(self._books)

File: practica_iterator/src/collection/BookCollection.java
package collection;

import iterator.*;
import model.Book;

import java.util.ArrayList;
import java.util.List;

/**
 * La clase BookCollection implementa la interfaz MyCollection para gestionar una colección de objetos Book.
 * Actúa como el "Agregado Concreto" en el patrón Iterator, ya que proporciona métodos
 * para crear diferentes tipos de iteradores para su colección interna de libros.
 */
public class BookCollection implements MyCollection<Book> {
    // La lista interna que almacena los objetos Book.
    // Esta es la colección sobre la cual los iteradores operarán.
    private final List<Book> books = new ArrayList<>();

    /**
     * Añade un nuevo libro a la colección.
     * @param item El libro a añadir.
     */
    @Override
    public void addItem(Book item) {
        books.add(item);
    }

    /**
     * Crea y devuelve un iterador para recorrer la colección de libros en orden normal.
     * Este es un "Método de Fábrica" para producir un "Iterador Concreto" (BookIterator).
     * @return Una instancia de BookIterator.
     */
    @Override
    public MyIterator<Book> createIterator() {
        return new BookIterator(books);
    }

    /**
     * Crea y devuelve un iterador para recorrer la colección de libros en orden inverso.
     * Este es otro "Método de Fábrica" para producir un "Iterador Concreto" (ReverseBookIterator).
     * @return Una instancia de ReverseBookIterator.
     */
    @Override
    public MyIterator<Book> createReverseIterator() {
        return new ReverseBookIterator(books);
    }

    /**
     * Crea y devuelve un iterador para recorrer la colección de libros ordenados alfabéticamente por autor.
     * Este es un "Método de Fábrica" para producir un "Iterador Concreto" (AuthorAlphabeticalBookIterator).
     * @return Una instancia de AuthorAlphabeticalBookIterator.
     */
    @Override
    public MyIterator<Book> createAuthorAlphabeticalIterator() {
        return new AuthorAlphabeticalBookIterator(books);
    }

    /**
     * Devuelve el número de libros en la colección.
     * @return El tamaño de la colección.
     */
    public int size() {
        return books.size();
    }

    /**
     * Método opcional para cargar datos de ejemplo en la colección de libros.
     */
    public void loadSampleData() {
        addItem(new Book("El Quijote", "Cervantes", "Novela", 1605));
        addItem(new Book("Cien Años de Soledad", "Gabriel García Márquez", "Realismo Mágico", 1967));
        addItem(new Book("La Sombra del Viento", "Carlos Ruiz Zafón", "Misterio", 2001));
    }
}